apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/PrunableArchiveMigrator.java

```diff
@@ -0,0 +1,93 @@
+        this.dirProvider = dirProvider;
+        this.blockchainConfig = blockchainConfig;
+        this.zip = zip;
+        this.databaseManager = databaseManager;
+        this.registry = registry;
+    }
+
+    public void migrate() {
+        String migrateOption = optionDAO.get(MIGRATE_OPTION);
+        if (migrateOption == null) {
+            String currentShard = optionDAO.get(CURRENT_SHARD_OPTION);
+            long shardIdForMigration = 1;
+            if (currentShard != null) {
+                shardIdForMigration = Long.parseLong(currentShard);
+            }
+            long finalShardIdForMigration = shardIdForMigration;
+            List<Shard> shardsForMigration = shardDao.getAllCompletedShards().stream().filter(shard -> shard.getShardId() >= finalShardIdForMigration).sorted(Comparator.comparing(Shard::getShardId)).collect(Collectors.toList());
+            List<String> tablesToExclude = registry.getDerivedTables().stream().filter(t -> t instanceof PrunableDbTable).map(DerivedTableInterface::getName).collect(Collectors.toList());
+            tablesToExclude.add(ShardConstants.DATA_TAG_TABLE_NAME);
+            for (Shard shard : shardsForMigration) {
+                try {
+                    optionDAO.set(CURRENT_SHARD_OPTION, String.valueOf(shard.getShardId()));
+                    Path tempDirectory = Files.createTempDirectory("prunable-shard-migration-" + shard.getShardId());
+                    ShardNameHelper shardNameHelper = new ShardNameHelper();
+                    Path shardArchivePath = dirProvider.getDataExportDir().resolve(shardNameHelper.getCoreShardArchiveNameByShardId(shard.getShardId(), blockchainConfig.getChain().getChainId()));
+
+                    String tempDirectoryString = tempDirectory.toAbsolutePath().toString();
+                    zip.extract(shardArchivePath.toAbsolutePath().toString(), tempDirectoryString);
+                    CsvExporter csvExporter = createExporter(tempDirectory);
+                    csvExporter.exportShardTableIgnoringLastZipHashes(shard.getShardHeight(), 100);
+                    String zipName = "shard-" + shard.getShardId() + ".zip";
+                    Path newArchive = tempDirectory.resolve(zipName);
+                    byte[] hash = zip.compressAndHash(newArchive.toAbsolutePath().toString(), tempDirectoryString, 0L, (dir, name) -> !tablesToExclude.contains(name.substring(0, name.indexOf(".csv"))), false);
+                    Files.move(newArchive, shardArchivePath, StandardCopyOption.REPLACE_EXISTING);
+                    shard.setCoreZipHash(hash);
+                    shard.setPrunableZipHash(new byte[32]); // not null to force prunable archive recreation
+                    shardDao.updateShard(shard);
+                    FileUtils.clearDirectorySilently(tempDirectory); // clean is not mandatory, but desirable
+                }
+                catch (IOException e) {
+                    throw new RuntimeException(e);
```

Comment:
What will happen if migration process is stopped in the middle ? Will it survive after that?

Reply:
Recovery will be recovered from not finished step.